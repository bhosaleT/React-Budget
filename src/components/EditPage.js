import React from "react";

const EditPage = (props) => {
    console.log(props);
    return (
        <div>
        Editing the expenses with id of {props.match.params.id}
        </div>
    )
};

export default EditPage;