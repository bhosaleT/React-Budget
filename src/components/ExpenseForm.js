import React, { Component } from "react";
import moment from "moment";
import { SingleDatePicker } from "react-dates";
import "react-dates/initialize";

// const date = new Date();

const now = moment();
console.log(now.format("MMM Do YYYY, h:mm a"));

export default class ExpenseForm extends Component {
  constructor(props) {
    super(props);

    this.state = {
      description: props.expense ? props.expense.description : "",
      note: props.expense ? props.expense.note : "",
      amount: props.expense ? (props.expense.amount / 100).toString() : "",
      createdAt: props.expense ? moment(props.expense.createdAt) : moment(),
      calFocused: false,
      error: ""
    };
  }

  onDescriptionChange = e => {
    const description = e.target.value;
    this.setState(() => ({ description }));
  };

  onChangeNote = e => {
    const note = e.target.value;
    this.setState(() => ({ note }));
  };

  onAmountChange = e => {
    const amount = e.target.value;
    //check if the amount is fitting a particular format.
    if (!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)) {
      this.setState(() => ({ amount }));
    }
  };

  onDateChange = createdAt => {
    if (createdAt) {
      this.setState(() => ({ createdAt }));
    }
  };

  onFocusChange = ({ focused }) => {
    this.setState(() => ({ calFocused: focused }));
  };

  onSubmit = e => {
    e.preventDefault();

    if (!this.state.description || !this.state.amount) {
      this.setState(() => ({ error: "Please add a title and amount" }));
    } else {
      this.setState(() => ({ error: "" }));
      this.props.onSubmit({
        description: this.state.description,
        amount: parseFloat(this.state.amount, 10) * 100,
        createdAt: this.state.createdAt.valueOf(),
        note: this.state.note
      });
    }
  };

  render() {
    return (
      <form className="form" onSubmit={this.onSubmit}>
        {this.state.error && <p className="form__error">{this.state.error}</p>}
        <input
          type="text"
          placeholder="Title"
          autoFocus
          value={this.state.description}
          className="text-input"
          onChange={this.onDescriptionChange}
        />
        <input
          type="text"
          value={this.state.amount}
          onChange={this.onAmountChange}
          className="text-input"
          placeholder="Amount"
        />
        <SingleDatePicker
          date={this.state.createdAt}
          onDateChange={this.onDateChange}
          focused={this.state.calFocused}
          onFocusChange={this.onFocusChange}
          numberOfMonths={1}
          isOutsideRange={() => false}
        />
        <textarea
          rows="4"
          cols="40"
          className="text-area"
          placeholder="Add a note for your expense (optional)"
          value={this.state.note}
          onChange={this.onChangeNote}
        />
        <div>
          <button className="button">Add Expense</button>
        </div>
      </form>
    );
  }
}
