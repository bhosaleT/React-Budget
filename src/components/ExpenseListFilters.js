import React from "react";
import { connect } from "react-redux";
import { setTextFilter } from "../actions/filters";
import { sortByDate, sortByAmount } from '../actions/filters'; 

/* ======================= WORKING OF DISPATCH WITH CONNECT FOR FILTERING WITH TEXT =====================
-- we use connect to connect the expeselistfilters component to the redux store.
-- for the input we add input of props.filters.text to make it show the search term , and onChange we want to change the text in and filter accordingly
-- the connect() function also gives us a dispatch method, this is the same dispatch method as that on the redux store.
-- so we use it using props.dispatch(in here we call the action function we already wrote i.e. setTextFilter(and to this we give the input of the text field))
*/

const ExpenseListFilters = props => (
  <div>
    <input
      type="text"
      value={props.filters.text}
      onChange={e => {
        props.dispatch(setTextFilter(e.target.value));
      }}
    />
    <select value={props.filters.sortBy} onChange = {(e)=>{
        props.dispatch(e.target.value === 'date' ? sortByDate() : sortByAmount() )
    }}>
     <option value="date">Date</option>
     <option value="amount">Amount</option>
    </select>
  </div>
);

const mapStateToProps = state => {
  return {
    filters: state.filters
  };
};

export default connect(mapStateToProps)(ExpenseListFilters);
