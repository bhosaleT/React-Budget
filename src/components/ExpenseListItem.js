import React from "react";
import { removeExpense } from "../actions/expenses";
import { connect } from "react-redux";

/*  ================= WORKING OF DELETE BUTTON =============
-- the mapStateToProps is not always required so to remove a state, we will import our action function removeExpense and call dispatch(removeExpense(with the expense id))
-- $ This code exists in this component because the id is already passed through props and destructured here so we already have it here to be directly passed into the          function.
-- $ Remember the action functions take in objects {}.
*/

const ExpenseListItem = ({ dispatch, id, description, amount, createdAt }) => {
  return (
    <div>
      <h3>{description}</h3>
      <p>Amount: {amount}</p>
      <p>Time: {createdAt}</p>
      <button
        onClick={() => {
          dispatch(removeExpense({ id }));
        }}
      >
        Remove
      </button>
    </div>
  );
};

export default connect()(ExpenseListItem);
