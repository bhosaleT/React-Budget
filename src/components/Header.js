import React from "react";
import { NavLink } from "react-router-dom";
import { startLogout } from "../actions/auth";
import { connect } from "react-redux";

export const Header = ({ startLogout }) => (
  <header>
    <h1>Pocket Money</h1>
    <NavLink to="/" activeClassName="is-active" exact={true}>
      Dashboard
    </NavLink>
    <NavLink to="/create" activeClassName="is-active">
      {" "}
      Create Expense{" "}
    </NavLink>
    <button onClick={startLogout}>Logout</button>
  </header>
);

const mapDispatchToProps = dispatch => ({
  startLogout: () => dispatch(startLogout())
});

export default connect(
  undefined,
  mapDispatchToProps
)(Header);
