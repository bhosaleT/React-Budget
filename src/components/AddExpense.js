import React from "react";
import ExpenseForm from "./ExpenseForm";
import { connect } from "react-redux";
import { startAddExpense } from "../actions/expenses";

export class AddExpensePage extends React.Component {
  onSubmit = expense => {
    this.props.startAddExpense(expense);
    this.props.history.push("/");
  };
  render() {
    return (
      <div>
        <h1>Add Expense</h1>
        <ExpenseForm onSubmit={this.onSubmit} />
      </div>
    );
  }
}

const mapDispatchToProps = (dispatch) => {
  return {
    startAddExpense: expense => dispatch(startAddExpense(expense))
  };
};

export default connect(
  undefined,
  mapDispatchToProps
)(AddExpensePage);

/* ============= Creating a more testable component =============== 
-- during the testing phase we would have had problems tesing the addExpense because its being passed in not as a prop but imported.
-- so instead we make a new function onSubmit and to it we pass the dispatch(addExpense(expense)).
-- this is much easier to test.
*/
