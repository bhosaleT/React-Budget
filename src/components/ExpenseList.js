import React from "react";
import { connect } from "react-redux";
import ExpenseListItem from "./ExpenseListItem";
import getVisibleExpenses from "../selectors/expenses";

export const ExpenseList = props => (
  <div className="content-container">
    <div className="list-header">
      <div className="show-for-mobile">Expenses</div>
      <div className="show-for-desktop">Expense</div>
      <div className="show-for-desktop">Amount</div>
    </div>

   <div className="list-body">
      {props.expenses.length === 0 ? (
        <div>
          <span className="list-item--message">No Expenses</span>
        </div>
      ) : (
          props.expenses.map(expense => {
            return <ExpenseListItem key={expense.id} {...expense} />;
            {
              /* to destructuring expenses at the ExpenseListItem we will add spreading to the props passing */
            }
          })
        )}
   </div>
  </div>
);

/* 
    ==================================== WORKING OF CONNECT =========================== 
- The connect is a named export from the package react-redux which lets us connect our react components to the redux store
- we use connect()(ExpanseList) and send back ExpanseList in it because it is a hoc which takes in a component and sends back
  a modified componet with some additions.
- we make another function which is called mapStateToProps this function takes the state from the store which is passed down from the App.js file and
  maps it to the props of the ExpanseList component.
- this can be used in props.expenses.length.
- Now our component can access all the information it requires form the store.
- As the store changes the component will change automatically.
- ********* ADDING FILTERING *********** we change the expenses props to take in getVisibleExpenses() function which takes in expenses and filters and sends back a filtered   list.
*/

const mapStateToProps = state => {
  return {
    expenses: getVisibleExpenses(state.expenses, state.filters)
  };
};

export default connect(mapStateToProps)(ExpenseList);
