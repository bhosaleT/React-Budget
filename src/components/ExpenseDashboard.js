import React from "react";
import ExpenseList from "./ExpenseList";
import ExpenseListFilters from "./ExpenseListFilters";
import ExpensesSummary from './ExpenseSummary';

const ExpenseDashboardPage = () => (
  <div>
    <ExpensesSummary />
    <ExpenseListFilters />
    <ExpenseList />
  </div>
);

export default ExpenseDashboardPage;
