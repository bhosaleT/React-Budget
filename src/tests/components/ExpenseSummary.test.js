import React from 'react';
import { shallow } from 'enzyme';
import { ExpensesSummary } from '../../components/ExpenseSummary';

//Test when we have a singular expense
test('should correctly render expenses summary with one expense', ()=>{
const wrapper = shallow(<ExpensesSummary expenseCount = {1} expensesTotal = {235}/>);
expect(wrapper).toMatchSnapshot();
})

test('should correctly render expenses summary with multiple expense', () => {
const wrapper = shallow(<ExpensesSummary expenseCount = {2} expensesTotal = {345} />);
expect(wrapper).toMatchSnapshot();
})