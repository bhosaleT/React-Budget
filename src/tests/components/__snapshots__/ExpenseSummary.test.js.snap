// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should correctly render expenses summary with multiple expense 1`] = `
<div>
  <h1>
    Viewing 
    2
     
    expenses
     totalling 
    $3.45
  </h1>
</div>
`;

exports[`should correctly render expenses summary with one expense 1`] = `
<div>
  <h1>
    Viewing 
    1
     
    expense
     totalling 
    $2.35
  </h1>
</div>
`;
